In [74]:
# pip install notebook

In [75]:
import pandas as pd
import numpy as np

In [76]:
df = pd.read_excel(r"E:\00-Study in Adelaide\Forage\02-BCGX\10-K data.xlsx")

In [77]:
print(df.columns)

Index(['Company', 'Year', 'Total Revenue', 'Net Income', 'Total Assets',
       'Total Liabilities', 'Cash Flow from Operating Activities'],
      dtype='object')


In [78]:
import pandas as pd
import numpy as np

def calculate_growth_rates(df):
    """
    Calculate year-over-year growth rates for key metrics
    """
    growth_df = df.copy()
    metrics = ['Total Revenue', 'Net Income', 'Total Assets', 
               'Total Liabilities', 'Cash Flow from Operating Activities']
    
    for metric in metrics:
        column_name = f'{metric.replace("Total ", "").replace(" ", "_")} Growth (%)'
        growth_df[column_name] = growth_df.groupby('Company')[metric].pct_change() * 100
    
    return growth_df

def calculate_financial_ratios(df):
    """
    Calculate key financial ratios and their changes
    """
    ratio_df = df.copy()
    
    # Profitability ratios
    ratio_df['Profit Margin'] = ratio_df['Net Income'] / ratio_df['Total Revenue'] * 100
    ratio_df['Profit Margin Change'] = ratio_df.groupby('Company')['Profit Margin'].pct_change()
    
    # Efficiency ratios
    ratio_df['Asset Turnover'] = ratio_df['Total Revenue'] / ratio_df['Total Assets']
    ratio_df['Asset Turnover Change'] = ratio_df.groupby('Company')['Asset Turnover'].pct_change() * 100
    
    # Leverage ratios
    ratio_df['Debt Ratio'] = ratio_df['Total Liabilities'] / ratio_df['Total Assets'] * 100
    ratio_df['Debt Ratio Change'] = ratio_df.groupby('Company')['Debt Ratio'].pct_change()
    
    # Cash flow ratios
    ratio_df['Operating Cash Flow Ratio'] = ratio_df['Cash Flow from Operating Activities'] / ratio_df['Total Revenue'] * 100
    ratio_df['Operating Cash Flow Ratio Change'] = ratio_df.groupby('Company')['Operating Cash Flow Ratio'].pct_change()
    
    return ratio_df

def analyze_companies(df):
    """
    Comprehensive financial analysis of companies
    """
    results = {}
    
    # Calculate all metrics first
    analyzed_df = calculate_growth_rates(df)
    analyzed_df = calculate_financial_ratios(analyzed_df)
    
    # 1. Annual Industry Trends
    results['industry_trends'] = analyzed_df.groupby('Year').agg({
        'Total Revenue': 'sum',
        'Net Income': 'sum',
        'Cash Flow from Operating Activities': 'sum'
    }).round(0)
    
    # 2. Company Comparison (now using calculated metrics)
    metrics_for_comparison = [
        'Revenue Growth (%)', 
        'Net Income Growth (%)',
        'Profit Margin', 
        'Asset Turnover',
        'Debt Ratio', 
        'Operating Cash Flow Ratio'
    ]
    
    # Ensure all required columns exist before aggregating
    available_metrics = [m for m in metrics_for_comparison if m in analyzed_df.columns]
    results['company_metrics'] = analyzed_df.groupby('Company')[available_metrics].agg([
        'mean', 'min', 'max'
    ]).round(2)
    
    # 3. Efficiency Analysis
    efficiency_metrics = ['Asset Turnover', 'Profit Margin', 'Operating Cash Flow Ratio']
    results['efficiency_analysis'] = analyzed_df.groupby('Company')[efficiency_metrics].mean().round(3)
    
    return analyzed_df, results

def display_results(df, results):
    """
    Display formatted analysis results
    """
    pd.set_option('display.float_format', lambda x: '%.2f' % x)
    
    print("\n=== Detailed Company Analysis ===")
    display_columns = [
        'Company', 'Year', 
        'Revenue Growth (%)', 'Net Income Growth (%)',
        'Profit Margin', 'Asset Turnover', 
        'Debt Ratio', 'Operating Cash Flow Ratio'
    ]
    
    # Only display columns that exist
    available_columns = [col for col in display_columns if col in df.columns]
    print(df[available_columns].sort_values(['Company', 'Year']))
    
    print("\n=== Industry Annual Trends ===")
    print(results['industry_trends'])
    
    print("\n=== Company Metrics Comparison ===")
    print(results['company_metrics'])
    
    print("\n=== Efficiency Analysis ===")
    print(results['efficiency_analysis'])

def main():
    # Load and analyze data
    df = pd.read_excel(r"E:\00-Study in Adelaide\Forage\02-BCGX\10-K data.xlsx")
    analyzed_df, results = analyze_companies(df)
    display_results(analyzed_df, results)

if __name__ == "__main__":
    main()


=== Detailed Company Analysis ===
     Company  Year  Revenue Growth (%)  Profit Margin  Asset Turnover  \
8      Apple  2022                2.88          25.31            1.12   
7      Apple  2023               -1.98          25.31            1.09   
6      Apple  2024                 NaN          23.97            1.07   
2  Microsoft  2022               -6.44          36.69            0.54   
1  Microsoft  2023              -13.55          34.15            0.51   
0  Microsoft  2024                 NaN          35.96            0.48   
5      Tesla  2021              -33.93          10.49            0.87   
4      Tesla  2022              -15.82          15.45            0.99   
3      Tesla  2023                 NaN          15.47            0.91   

   Debt Ratio  Operating Cash Flow Ratio  
8       85.64                      30.98  
7       82.37                      28.84  
6       84.40                      30.24  
2       54.35                      44.91  
1       49.94      

KEY FINDINGS

1. Growth Performance
   - Apple shows the most stable growth pattern with revenue fluctuation between -1.98% to 2.88%
   - Microsoft experienced declining growth trends, with revenue growth ranging from -13.55% to -6.44%
   - Tesla demonstrates high growth volatility (-33.93% to -15.82%), indicating market challenges
   
2. Profitability Analysis
   - Microsoft leads in profitability with the highest average profit margin (35.60%)
   - Apple maintains consistent profit margins around 24.86%
   - Tesla shows improving but lower profit margins, averaging 13.80%

3. Operational Efficiency
   - Apple leads in asset efficiency with the highest asset turnover (1.09x)
   - Tesla shows moderate asset utilization (0.92x)
   - Microsoft has lower asset turnover (0.51x) but compensates with higher margins

4. Financial Structure
   - Apple has the highest debt ratio (82-85%), indicating higher financial leverage
   - Microsoft maintains moderate leverage (47-54%)
   - Tesla shows the lowest debt ratio (40-49%), suggesting conservative financing

5. Cash Flow Efficiency
   - Microsoft demonstrates strongest cash flow efficiency (44.87% operating cash flow ratio)
   - Apple maintains stable cash flows (30.02%)
   - Tesla shows lower but improving cash flow ratios (17.71%)

6. Competitive Position
   - Microsoft: High profitability focus with strong cash generation
   - Apple: Balanced performance with high operational efficiency
   - Tesla: Growth-focused with improving fundamentals

7. Risk Profile
   - Apple: Moderate risk with high leverage but stable operations
   - Microsoft: Lower risk with strong margins and moderate leverage
   - Tesla: Higher operational risk but lower financial risk